pkg/v2/tkr/resolver/internal/resolver_impl.go

```diff
@@ -0,0 +1,389 @@
+	if !shouldSet {
+		delete(ls, label)
+		return
+	}
+	ls[label] = ""
+}
+
+func (cache *cache) shippedOSImages(tkr *runv1.TanzuKubernetesRelease) data.OSImages {
+	osImages := make(data.OSImages, len(tkr.Spec.OSImages))
+	for _, osImageRef := range tkr.Spec.OSImages {
+		osImages[osImageRef.Name] = cache.osImages[osImageRef.Name] // nil if OSImage hasn't been added yet
+	}
+	return osImages
+}
+
+// Pre-reqs: tkr is NEVER nil
+func (cache *cache) addToOSImageToTKRs(osImages data.OSImages, tkr *runv1.TanzuKubernetesRelease) {
+	for osImageName := range osImages { // we only need name, value MAY be nil (but we still want the name)
+		tkrs, exists := cache.osImageToTKRs[osImageName]
+		if !exists {
+			tkrs = make(data.TKRs, 1) // an OSImage is shipped by (at least) 1 TKR
+			cache.osImageToTKRs[osImageName] = tkrs
+		}
+		tkrs[tkr.Name] = tkr
+	}
+}
+
+// populate cache.osImageToTKRs and cache.tkrToOSImages
+// Pre-reqs: osImage is NEVER nil
+func (cache *cache) addOSImage(osImage *runv1.OSImage) {
+	cache.augmentOSImage(osImage)
+
+	cache.osImages[osImage.Name] = osImage
+
+	tkrs, exists := cache.osImageToTKRs[osImage.Name]
+	if !exists {
+		tkrs = data.TKRs{}
+		cache.osImageToTKRs[osImage.Name] = tkrs
+	}
+	cache.addToTKRToOSImages(tkrs, osImage)
+}
+
+// augmentOSImage:
+// - sets missing k8s-version-prefix labels
+// - sets missing os-* labels
+// - sets image-type label
+// - set <image-type>-<ref-field> labels
+// - sets/removes incompatible, invalid labels based on status conditions
+func (cache *cache) augmentOSImage(osImage *runv1.OSImage) {
+	osImage.Labels = labels.Merge(osImage.Labels, version.Prefixes(version.Label(osImage.Spec.KubernetesVersion)))
+
+	osImage.Labels[runv1.LabelOSType] = osImage.Spec.OS.Type
+	osImage.Labels[runv1.LabelOSName] = osImage.Spec.OS.Name
+	osImage.Labels[runv1.LabelOSVersion] = osImage.Spec.OS.Version
+	osImage.Labels[runv1.LabelOSArch] = osImage.Spec.OS.Arch
+
+	imageType := osImage.Spec.Image.Type
+	osImage.Labels[runv1.LabelImageType] = imageType
+
+	setRefLabels(osImage.Labels, imageType, osImage.Spec.Image.Ref)
+
+	ensureLabel(osImage.Labels, runv1.LabelIncompatible, conditions.IsFalse(osImage, runv1.ConditionCompatible))
+	ensureLabel(osImage.Labels, runv1.LabelInvalid, conditions.IsFalse(osImage, runv1.ConditionValid))
+}
+
+func setRefLabels(ls labels.Set, prefix string, ref map[string]interface{}) {
+	for name, value := range ref {
+		prefixedName := prefix + "-" + name
+		if value, ok := value.(map[string]interface{}); ok {
+			setRefLabels(ls, prefixedName, value)
+			continue
+		}
+		ls[prefixedName] = fmt.Sprint(value)
+	}
+}
+
+// Pre-reqs: osImage is NEVER nil
+func (cache *cache) addToTKRToOSImages(tkrs data.TKRs, osImage *runv1.OSImage) {
+	for tkrName := range tkrs {
+		cache.tkrToOSImages[tkrName][osImage.Name] = osImage // cache.osImagesShippedByTKR[tkrName] is NEVER nil
+	}
+}
+
+func normalize(query data.Query) data.Query {
```

Comment:
nit: a comment about what this function would help

Reply:
Done